service/s3/unmarshal.go

```diff
@@ -0,0 +1,111 @@
+package s3
+
+import (
+	"bytes"
+	"encoding/xml"
+	"io"
+	"net/http"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/awserr"
+	restlegacy "github.com/aws/aws-sdk-go-v2/private/protocol/rest"
+)
+
+// protoGetObjectUnmarshaler defines unmarshaler for ProtoGetObject operation
+type protoGetObjectUnmarshaler struct {
+	output *GetObjectOutput
+}
+
+// unmarshalOperation is the top level method used with a handler stack to unmarshal an operation response
+// This method calls appropriate unmarshal shape functions as per the output shape and protocol used by the
+// service.
+func (u protoGetObjectUnmarshaler) unmarshalOperation(r *aws.Request) {
+	var buff bytes.Buffer
+	body := io.TeeReader(r.HTTPResponse.Body, &buff)
+	// an xml decoder should be used to decode an xml document
+	d := xml.NewDecoder(body)
+	startToken, err := d.Token()
+	if err != nil && err != io.EOF {
```

Comment:
Attempting to get the token from a decoder if there is not response body will result in an error. This can happen if the request fails for connection error, or there just is no XML .

Specifically for this operation, the successful response body probably won't contain a token. The prototype should first check if it thinks the operation was successful or not based on status code. If the operation was not successful, then the unmarshal error could be called with the first token.

We probably don't want to optimize for the 200 error case. But use that as a specific customization.

Reply:
```go 
startToken, err:= d.Token()
if err!= nil && err != io.EOF {
	r.Error = err
	return
}

// isRequestError checks if operation response returned a error response
if isRequestError(r) {
	r.Error = unmarshalErrorPrototype(r, d, startToken)
	return
}
```

When a response body doesn't have xml body, we should get an err of io.EOF, and proceed to check if isRequestError. 

My intial thought was that if operation is one of the few operations where we need to check for status 200 error, we should generate a different isRequestError definition, where we should also have the start tag as an param. Thus we try to retrieve the startToken before checking for error.

```go 
func isRequestError(r *aws.Request, startToken xml.Token)
``` 

The current implementation won't address the case where HTTPResponse.Body is nil. We should probably add a check for if response body is nil ans skip trying to retrieve the start tag.